dspace-server-webapp/src/main/java/org/dspace/app/rest/security/SubscriptionSecurityBean.java

```diff
@@ -0,0 +1,75 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.security;
+
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+
+import org.dspace.app.rest.utils.DSpaceObjectUtils;
+import org.dspace.authorize.service.AuthorizeService;
+import org.dspace.content.DSpaceObject;
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+import org.dspace.eperson.EPerson;
+import org.dspace.eperson.service.EPersonService;
+import org.dspace.services.RequestService;
+import org.dspace.utils.DSpace;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Methods of this class are used on PreAuthorize annotations
+ * to check security on subscriptions endpoint
+ * 
+ * @author Mykhaylo Boychuk (mykhaylo.boychuk at 4science.it)
+ */
+@Component(value = "subscriptionSecurity")
+public class SubscriptionSecurityBean {
+
+    @Autowired
+    private EPersonService ePersonService;
+    @Autowired
+    private AuthorizeService authorizeService;
+    @Autowired
+    private DSpaceObjectUtils dspaceObjectUtil;
+
+    public boolean isEnabelToCreateSubscription(Context context) throws SQLException {
```

Comment:
@tdonohue  We can't use `hasPermission(#id, 'resource', 'READ')` because the createAndReturn(Context context) method
https://github.com/DSpace/DSpace/pull/8620/files#diff-7b083a954bbeb1f357856c1cb0872ab5bec3df518f0f64a5c7727ed1bc44fc2bR147
doesn't contains the params `id` & `resource`, we retrieve them internaly from the request.

Reply:
@Micheleboychuk : There is a createAndReturn method that takes in a UUID on `DSpaceRestRepository` which is often used in this scenario: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java#L318

See how Community & Collection Rest Repositories implement this same concept of creating a new object which is "linked" to something else:
https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java#L314
https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java#L104

I see this as a similar situation... you are creating a new Subscription which is linked to a Resource.  So, I don't see why we couldn't just do this easier like this in SubscriptionRestRepository:
```
@Override
@PreAuthorize("hasPermission(#id, 'resource', 'READ') && hasAuthority('AUTHENTICATED')")
protected SubscriptionRest createAndReturn(Context context, UUID id) throws SQLException, AuthorizeException {
...
}
```

I just don't want us to create (seemingly unnecessary) new PreAuthorize statements when we can use what already exists in the codebase similar to other `*RestRepository` classes

Reply:
@tdonohue thanks for feedback
approach you suggest doesn't take into account the following use case:
- An administrator can create subscriptions for other users.
so we also need the eperson_id attribute in the security plugin.

Reply:
@Micheleboychuk : I'm pretty sure it would work fine.  To clarify, I'm asking for something like this (notice the "(NEW)" if statement):

```
@Override
@PreAuthorize("hasPermission(#id, 'resource', 'READ') && hasAuthority('AUTHENTICATED')")
protected SubscriptionRest createAndReturn(Context context, UUID id) throws SQLException, AuthorizeException {
        HttpServletRequest req = getRequestService().getCurrentRequest().getHttpServletRequest();
        String epersonId = req.getParameter("eperson_id");
        
        if (StringUtils.isBlank(id) || StringUtils.isBlank(epersonId)) {
            throw new UnprocessableEntityException("Both eperson than DSpaceObject uuids must be provided!");
        }

        try {
            DSpaceObject dSpaceObject = dspaceObjectUtil.findDSpaceObject(context, UUID.fromString(id));
            EPerson ePerson = ePersonService.findByIdOrLegacyId(context, epersonId);
            if (Objects.isNull(ePerson) || Objects.isNull(dSpaceObject)) {
                throw new DSpaceBadRequestException("Id of person or dspace object must represents reals ids");
            }

           // (NEW) Verify that EPerson matches current user or current user is Admin
           if ((ePerson != context.getCurrentUser) && !authorizeService.isAdmin(context)) {
              throw new AuthorizeException("Only admins can create subscriptions for other users");
           }

           ... rest of code ...
}
```

Basically, this seems like the *first situation* where we've decided we need to move all the authorization checks into a custom security-checking bean.  I don't think that's necessary, as we can just do the extra security check in the method itself.

The reason I prefer this approach is that it keeps all the code in one place (in SubscriptionRestRepository), instead of having to go find what "@subscriptionSecurity" references in order to determine what permissions it is checking.

Reply:
We checked and it is not possible to use the `createAndReturn` method with the parent parameter because it would require us to use "parent" as name of the request parameter and of course this would make no sense for this use case.

Personally I would recommend to encapsulate all the security logic in a dedicated bean implementation as was proposed by @Micheleboychuk  but as you suggest as an alternative to keep the logic in the repository method body we will update the PR as by your suggestion